### ISSUE TITLE
Possible issue with notification's "data" column re-assignment.

### ISSUE THREAD

**Issue by Infotaku** [rpush/rpush#156]
Hello everyone !

I'm currently in the process of developing a Cequel client for rpush and I've just stumbled across a portion of code I cannot stop to stare at : https://github.com/rpush/rpush/blob/master/lib/rpush/client/active_record/notification.rb#L22-L26

        def data=(attrs)
          return unless attrs
          fail ArgumentError, "must be a Hash" unless attrs.is_a?(Hash)
          write_attribute(:data, multi_json_dump(attrs.merge(data || {})))
        end

My issue is with the last line of that method. It seems to me like 

        attrs.merge(data || {})

should actually be 

        (data || {}).merge(attrs)

otherwise, the current value for a given key would always be preferred to the newly requested value for that same key, no ?
In my experience :

        new_attrs_data = {test: 1}
        current_data = {test: 2}
        new_attrs_data.merge(current_data) => {:test=>2}

But maybe this is the behaviour requested by some part of rpush implementation I did not quiet get.
Or maybe you are using a monkey-patch of some sort, or a gem that overwrites that merge behaviour.

So, bug or feature ? Can someone help me figure this out ?

**Comment by Infotaku** [comment]
Now that I think about it 

     (data || {}).merge(attrs)

is kinda weird in its own way. It would mean that any existing key could never be removed from the data hash.
So maybe the line should simply be 

     write_attribute(:data, multi_json_dump(attrs))

I'm eager to be hearing from you guys, that question is killing me :p

**Comment by ileitch** [comment]
Yeah, merge is kinda weird here. I don't remember why it's that way. Maybe there's a spec that may give us a clue?

**Comment by Infotaku** [comment]
I think I found the reason.

Somewhere in the data hash you seem to include some internal rpush keys like "__rpush_content_available__" that you use later to set APNS content_available property to 1.

I'm guessing the conservative merge is there to prevent that key for being erased by the setter.

Can you confirm ?

**Comment by pauldub** [comment]
I also think that the way this is done is weird and misleading. I think that preventing those key from being erased by the setter is a poor excuse for a broken setter.

Also, there is an error for the `notification=` setter which merges with `data` here: https://github.com/rpush/rpush/blob/master/lib/rpush/client/active_record/notification.rb#L31 Because of this a mixed `data` and `notification` FCM notification cannot be sent.

Would you accept a PR addressing both issues?

**Comment by aried3r** [comment]
Hey @pauldub, sure, if there is an actual error which can be shown with a test that would be great to have!

I haven't faced this error myself although we use mixed `data`/`notification` FCM notifications, but I'm definitely interested :)

**Comment by pauldub** [comment]
I'll look into the issues we had regarding mixed FCM notifications, it might be more specific. Regarding the `merge` behaviour you can easily see it here: 

```ruby
[3] pry(main)>  Rpush::Gcm::Notification.new(data: { foo: 1 }, notification: { title: 'bar' }).notification == { 'title' => 'bar', 'foo' => 1 }
=> true
```

It also depends on the order of assignment, if data is assigned last then it would not trigger that behaviour eg:

```ruby
[7] pry(main)> notification = Rpush::Gcm::Notification.new
[8] pry(main)> notification.notification = { title: 'foo' }
=> {:title=>"foo"}
[9] pry(main)> notification.data = { foo: 1 }
=> {:foo=>1}
[10] pry(main)> notification.notification
=> {"title"=>"foo"}
```

**Comment by aried3r** [comment]
> It also depends on the order of assignment, if data is assigned last then it would not trigger that behaviour

Ouch. I'd love to see this fixed, that's really not good. If you're up for it, I'd love to see a PR and get it merged if everything's ok. :)

**Comment by pauldub** [comment]
@aried3r could you give a look at https://github.com/tymate/rpush/commit/69bab9974c71c8c526b34ea4c70f789912f8e74c and the comments there please?
